include/odp/api/spec/stash.h

```diff
@@ -0,0 +1,288 @@
+	 *  calls.
+	 */
+	ODP_STASH_TYPE_FIFO
+
+} odp_stash_type_t;
+
+/**
+ * Stash operation mode
+ */
+typedef enum odp_stash_op_mode_t {
+	/** Multi-thread safe operation
+	 *
+	 *  Multiple threads operate on the stash. A stash operation
+	 *  (odp_stash_put() or odp_stash_get()) may be performed concurrently
+	 *  from multiple threads.
+	 */
+	ODP_STASH_OP_MT = 0,
+
+	/** Single thread operation
+	 *
+	 *  Multiple threads operate on the stash, but application ensures that
+	 *  a stash operation (odp_stash_put() or odp_stash_get()) is not
+	 *  performed concurrently from multiple threads.
+	 */
+	ODP_STASH_OP_ST,
+
+	/** Thread local operation
+	 *
+	 *  Only a single thread operates on the stash. Both stash operations
+	 *  (odp_stash_put() and odp_stash_get()) are always performed from the
+	 *  same thread.
+	 */
+	ODP_STASH_OP_LOCAL
+
+} odp_stash_op_mode_t;
+
+/**
+ * Stash capabilities
+ */
+typedef struct odp_stash_capability_t {
+	/** Maximum number of stashes of any type */
+	uint32_t max_stashes_any_type;
+
+	/** Maximum number of stashes of this type */
+	uint32_t max_stashes;
+
+	/** Maximum number of objects per stash
+	 *
+	 *  The value of zero means that limited only by the available
+	 *  memory size. */
+	uint64_t max_num_obj;
+
+	/** Maximum supported object handle size in bytes
+	 *
+	 *  At least 4 byte object size is always supported.
+	 */
+	uint32_t max_obj_size;
+
+} odp_stash_capability_t;
+
+/**
+ * Stash parameters
+ */
+typedef struct odp_stash_param_t {
+	/** Stash type
+	 *
+	 *  Select type of the stash to be created. The default value is
+	 *  ODP_STASH_TYPE_DEFAULT. Use capability to check if additional types
+	 *  are supported.
+	 */
+	odp_stash_type_t type;
+
+	/** Put operation mode
+	 *
+	 *  The default value is ODP_STASH_OP_MT. Usage of ODP_STASH_OP_ST or
+	 *  ODP_STASH_OP_LOCAL mode may improve performance when applicable.
+	 *  If ODP_STASH_OP_LOCAL is used, it must be set to both put_mode and
+	 *  get_mode.
+	 */
+	odp_stash_op_mode_t put_mode;
+
+	/** Get operation mode
+	 *
+	 *  The default value is ODP_STASH_OP_MT. Usage of ODP_STASH_OP_ST or
+	 *  ODP_STASH_OP_LOCAL mode may improve performance when applicable.
+	 *  If ODP_STASH_OP_LOCAL is used, it must be set to both put_mode and
+	 *  get_mode.
+	 */
+	odp_stash_op_mode_t get_mode;
+
+	/** Maximum number of objects
+	 *
+	 *  This is the maximum number of object handles application will store
+	 *  to the stash. The value must not exceed 'max_num_obj' capability.
+	 */
+	uint64_t num_obj;
+
+	/** Object handle size in bytes
+	 *
+	 *  Application uses this size object handles with put and get
+	 *  operations. Valid values are powers of two (1, 2, 4, 8, ... bytes)
+	 *  and must not exceed 'max_obj_size' capability.
+	 */
+	uint32_t obj_size;
+
+} odp_stash_param_t;
+
+/**
+ * Query stash capabilities
+ *
+ * Outputs capabilities of the given stash type on success.
+ * The default stash type (ODP_STASH_TYPE_DEFAULT) is always supported.
```

Comment:
Changing to this (max_stashes == 0:  type known but not supported):

 * Query stash capabilities
 *
 * Outputs capabilities of the given stash type on success. The stash type
 * is not supported if 'max_stashes' capability is zero. The default stash
 * type (ODP_STASH_TYPE_DEFAULT) is always supported. The function
 * returns failure if the given stash type is unknown to the implementation.

The idea is that a conforming ODP implementation would implement at least one default type stash. Max size of the stash may be small (down to single uint32_t) if resources are limited.

In general, basic features of all ODP APIs must be implemented by all implementations. It would be hard to write a portable application, if basic features may be missing (lot's of work around code needed).

Reply:
Ack.